src/Workspaces/Core/Portable/Options/GlobalOptionService.cs

```diff
@@ -91,12 +99,52 @@ static ImmutableHashSet<IOption> ComputeSerializableOptionsFromProviders(Immutab
             }
         }
 
+        private ImmutableArray<IOptionPersister> GetOptionPersisters()
+        {
+            // If _lazyOptionSerializers is already initialized, return its value directly.
+            if (_lazyOptionSerializers is { IsDefault: false } optionSerializers)
+            {
+                return optionSerializers;
+            }
+
+            ImmutableInterlocked.InterlockedInitialize(
+                ref _lazyOptionSerializers,
+                GetOptionPersistersSlow(_workspaceThreadingService, _optionSerializerProviders, CancellationToken.None));
+
+            return _lazyOptionSerializers;
+
+            // Local functions
+            static ImmutableArray<IOptionPersister> GetOptionPersistersSlow(
+                IWorkspaceThreadingService? workspaceThreadingService,
+                ImmutableArray<Lazy<IOptionPersisterProvider>> optionSerializerProviders,
+                CancellationToken cancellationToken)
+            {
+                if (workspaceThreadingService is not null)
+                {
+                    return workspaceThreadingService.Run(() => GetOptionPersistersAsync(optionSerializerProviders, cancellationToken));
+                }
+                else
+                {
+                    return GetOptionPersistersAsync(optionSerializerProviders, cancellationToken).WaitAndGetResult_CanCallOnBackground(cancellationToken);
+                }
```

Comment:
instead of getting the service, checking for null, and falling back to WaitAndGetResult here, should we instead export a DefaultWorkspaceThreadingService that has this as its implementation for .Run

Reply:
This is not a workspace service, so we can only export one total instance of it (MEF layering is specific to our workspace and language services).

Reply:
A two-layer indirection alternative exists, which I can implement separately to see if it's preferred.